pkg/flag/flag.go

```diff
@@ -0,0 +1,67 @@
+package flag
+
+import (
+	"crypto/tls"
+	"flag"
+
+	atlastls "github.com/infobloxopen/atlas-app-toolkit/pkg/tls"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+type GRPCFlags struct {
+	addr *string
+}
+
+type TLSFlags struct {
+	cert *string
+	key *string
+	ca *string
+}
+
+func HealthFlags() (*string, *string) {
+	addr := flag.String("health-addr", ":8080", "address to use for the health endpoint")
+	path := flag.String("health-path", "/health", "path to use for the health endpoint")
+	return addr, path
+}
+
+func MetricsFlags() (*string, *string) {
+	addr := flag.String("metrics-addr", ":9500", "address to use for the metrics endpoint")
+	path := flag.String("metrics-path", "/metrics", "path to use for the metrics endpoint")
+	return addr, path
+}
+
+func NewGRPCFlags() *GRPCFlags {
+	f := &GRPCFlags{}
+	f.addr = flag.String("addr", ":9000", "address to use for the gRPC endpoint")
+	return f
+}
+
+func (f *GRPCFlags) Addr() string {
+	return *f.addr
+}
+
+func NewTLSFlags() *TLSFlags {
+	f := &TLSFlags{}
+	f.cert = flag.String("cert", "", "path to the server certificate in PEM format")
+	f.key = flag.String("key", "", "path to the server private key in PEM format")
+	f.ca = flag.String("ca", "", "path to the CA PEM for validating the client cert; system CAs will be used if blank")
+
+	return f
+}
+
+func (f *TLSFlags) TLSConfig() (*tls.Config, error) {
+	return atlastls.NewTLSConfig(*f.cert, *f.key, *f.ca)
+}
+
+func (f *TLSFlags) WithGRPCTLSCreds() (grpc.ServerOption, error) {
```

Comment:
If we stick to this design, the resulting code from user side will be something like:
``` golang
tlsFlags := NewTLSFlags()
optWithTLS, err := tlsFlags.WithGRPCTLSCreds()
if err != nil {
    panic(err)
}
srv := grpc.NewServer(optWithTLS)
```
This leads us to a predefined set of flags for accepting TLS certs and keys. I see no reason to move TLS config initialization into toolkit. It can be googled in 5 seconds and implemented using grpc only code without any restrictions on keys/certs falg names/paths or anything else:
``` golang
credentials.NewServerTLSFromFile("cert/server.crt", "cert/server.key")
if err != nil {
    log.Fatalf("could not load TLS keys: %s", err)
}
grpcServer := grpc.NewServer(grpc.Creds(creds))
```

Reply:
We want a consistent interface to all apps built with the toolkit. That includes flags - there is another story being worked to allow consistent configuration via flags, env vars, and file. Without that, you are hard-coding the paths to the certs, which limits deployment flexibility substantially.

Reply:
It's not about hard-coding paths to certificates. Our flags(for tls config) currently limiting user from having any alternative type of input of tls-related things. Moreover, we want him to use our code for creating grpc.ServerOption for his grpc server which can be created only by him and only inside his code with any grpc.ServerOptions he wants.

Reply:
We want the consistent flags. That specific flag mechanism needs to be expanded into the general config mechanism that uses any of flag/env var/file. That should move to some other pkg for that purpose.

Nothing actually forces the author to use this - they can still do it exactly how you did above, if they want different flags or whatever for some reason. It's just a convenience.